### DISCUSSION TITLE
Open ended state machine diagram

### DISCUSSION THREAD

**Discussion by smol-ninja** [sablier-labs/flow#70] [Show and tell]
I hope it helps to visualize the open-ended streams as per the current implementation. if you think there could be some improvements in the diagram, please let me know and I will be happy to make those changes.

[Click here to view in Stately Studio](https://stately.ai/registry/editor/embed/dbf93aad-910f-435e-a5ad-297466483e19?mode=Design&machineId=c47a00c5-ff18-4087-a16c-5a6ac5c3d453)

- The rectangular boxes represent the states.
- The elliptical boxes represent the functions that can be called during a particular state.

<img width="1248" alt="Screenshot 2024-05-09 at 11 04 53" src="https://github.com/sablier-labs/v2-open-ended/assets/6676622/a0ddc931-001b-44a3-90c1-8bcb1517be44">

**Comment by andreivladbrg** [comment]
Thank you Shub for the diagram, since NFT PR has changed how things work I have created a new version of diagrams (they include `cancel` --> `pause` change):

### State diagram

![OE_state_diagram_correct](https://github.com/sablier-labs/v2-open-ended/assets/99738872/1f74de53-4fdc-4bd6-83e7-d0ab872d0f19)

---

### Calculation diagrams

![Amounts_Calculation](https://github.com/sablier-labs/v2-open-ended/assets/99738872/cede8066-3c8b-4064-8aad-a38b95eff72a)


--- 

I’ve put effort into these and would appreciate your feedback. Tagging @sablier-labs/solidity @razgraf

Drive links:
- State: https://drive.google.com/file/d/1C9gvhF0OtEsRY6Q68c4Wc4JUUR4TcxS0/view?usp=drive_link
- Amount calculations: https://drive.google.com/file/d/14VU6elpmk2CbpjD1SEAhRt3YFBnmiqjC/view?usp=drive_link

**Reply by smol-ninja** [reply]
Thanks for the new diagrams Andrei. It explains the specs very nicely. And I love the new specs of Open-ended tbh.

A few questions that come to my mind without having looked at the recent PR:

1. Do **ADJUST_RPS** and **PAUSE** trigger transfers to the sender and recipients, or do they only update values in storage?
2. I think **ADJUST_RPS** would also include `update_rps`?
3. The branch of `isPaused == true` and `debt > 0` is missing. It would return `bal` and not `ra`.
4. In the calculation diagram, if `debt > 0`, shouldn't the withdrawable equal `bal`?

For aesthetic reasons, we can re-draw them in Stately before putting them up in the docs. wdyt?

**Reply by andreivladbrg** [reply]
> Do ADJUST_RPS and PAUSE trigger transfers to the sender and recipients, or do they only update values in storage?

Nope, they don't transfer any token. Notice the red table (functions which call `ERC20.transfer`):

i've also update the arrows to have the color red to point out the fact that there is a `ERC20` transfer happening

<img width="183" alt="image" src="https://github.com/sablier-labs/v2-open-ended/assets/99738872/8474b94f-7b8a-47ef-9c1f-74b90aa2ddce">

> I think ADJUST_RPS would also include update_rps?

yes, thank you, missed this

> The branch of isPaused == true and debt > 0 is missing. It would return bal and not ra.

nope, this is not missing, after the stream is paused, the only amount available to be withdrawn is the remaining amount, because in `pause` function if there is debt, the balance will set to 0

also, there can't be "debt" when the stream is paused, since the  _if isPaused == true then rps == 0_
and, atm we don't track any "debt before pause" amount. but, as we decided, we will have a method: `payDebtAndPause`

> In the calculation diagram, if debt > 0, shouldn't the withdrawable equal bal?

nope. imagine the case where a stream has the rps adjusted, then ra is modified with the available amount at that moment in time. then more time passes, and the new balance (with the subtraction from the adjustRPS function) becomes insufficient for the streamed tokens, but the ra is still available to be withdrawn. (after you review the NFT PR, this will make more sense.)
please, lmk if i made myself clear.

---

> For aesthetic reasons, we can re-draw them in Stately before putting them up in the docs. wdyt

this may be subjective, but imo, Stately doesn't look better - but on the contrary, the new versions with colors help a lot to understand how the protocol works, and they are aesthetically better 
also, with Stately, you don't the same level of details as you do in app.diagrams.net

but let's give @razgraf the option to decide, since he is the master of design

**Reply by smol-ninja** [reply]
> nope, this is not missing, after the stream is paused, the only amount available to be withdrawn is the remaining amount, because in pause function if there is debt, the balance will set to 0

Got it. So `ra` is keeping track of debt too. Right?

> also, there can't be "debt" when the stream is paused

My question was about debt accumulated before `paused()` is called and how is that being carry forwarded. Now I think `ra` carries that.

> please, lmk if i made myself clear

Yes I think I understood, depending on your answer to my above question.

> but let's give @razgraf the option to decide, since he is the master of design

Always.

**Reply by andreivladbrg** [reply]
> Got it. So ra is keeping track of debt too. Right?

nope, it isn't keeping track of the debt. if the stream is paused with a balance of 0 and a debt of 100 tokens, the debt will "lost" and the remaining amount will still be 0

remainingAmount is keeping track of the tokens already streamed in case any action that effects the rps is taken

see also the thread here: https://github.com/sablier-labs/v2-open-ended/discussions/34#discussioncomment-9253880

**Reply by smol-ninja** [reply]
Got it. Thank you for explaining. In [this thread](https://github.com/sablier-labs/v2-open-ended/discussions/34#discussioncomment-9253880), I did not understand what you were saying so we softly agreed upon an implementation to understand how it works. Now I think I understand. 

I don't like this approach of "losing" debt upon pausing. So I have an idea, how about `pause` does not lose track of already accumulated debt? Separately there can be a different function called `void` (as mentioned [here](https://github.com/sablier-labs/v2-open-ended/issues/61) which can be called recipient) which resets debt to 0. I don't know about `void` but `pause` seems to be an important function for users.

I can clearly see a use case where someone might want to pause a stream without resetting the debt because they intend to pay it in the future but can no longer afford the accumulation of debt. This is just my hunch and can only be validated by talking to users. 

I'd also like to know why do you favour resetting debt upon `pausing`?

Started discussion on this: https://github.com/sablier-labs/v2-open-ended/discussions/86

**Reply by andreivladbrg** [reply]
@smol-ninja, could I kindly ask you to open a separate discussion for this? It seems like it’s not related to the state diagram itself, but it is an interesting idea for a new feature.

**Reply by smol-ninja** [reply]
OK I will.

**Reply by razgraf** [reply]
> the new versions with colors help a lot to understand how the protocol works

Completely agree that the colors help a lot with understanding the flow. Unfortunately, the design is very... Windows 95-ish 😅 Stately looks stylish, so maybe we can use the color coding from Andrei's version for a new state diagram? Looks like[ it should be possible](https://stately.ai/blog/2023-07-24-improving-your-flows-with-color).

**Reply by andreivladbrg** [reply]
Ok. Will make a new version as I am wokring on the cancel -> pause PR and the diagrams will change.

**Reply by razgraf** [reply]
Thank you! I've also checked the diagrams, but will wait for the final version to perform a more thorough read. In any case, great job with them both of you!

**Reply by PaulRBerg** [reply]
Yeah nice work with the diagram - this should definitely also goes in the docs once we get to that stage.

I will review later once it's updated.

**Comment by andreivladbrg** [comment]
@razgraf @PaulRBerg @smol-ninja  just wanted to let you know that I've updated the `.drawio` diagrams

And I've also created a `mermaid` version. Please let me know which version you prefer ( I am also working on a Stately version but it's WIP) After we have all three, we can decide.

(Note: sometimes mermaid doesn't work inside `details` block, please see gist below)

**Later edit:** After seeing the mermaid diagrams in dark mode, I noticed that the colors are not well chosen. I will edit them tomorrow

<details><summary>Mermaid diagrams</summary>
<p>

```mermaid
flowchart LR
    stream[(Stream Internal State)]:::green
    bal([Balance - bal]):::green
    rps([RatePerSecond - rps]):::green
    ra([RemainingAmount - ra]):::green
    ltu([Last Time Update - ltu]):::green

    stream --> bal
    stream --> rps
    stream --> ra
    stream --> ltu

    classDef green fill:#32cd32,stroke:#333,stroke-width:2px;
```

```mermaid
flowchart LR
    erc_transfers[(ERC20 Transfer Actions)]:::red
    dep([Deposit - add]):::red
    ref([Refund - extract]):::red
    wtd([Withdraw - extract]):::red

    erc_transfers --> dep
    erc_transfers --> ref
    erc_transfers --> wtd

    classDef red fill:#ff4e4e,stroke:#333,stroke-width:2px;
```

---

### State diagram

**Notes:**

1. The "update" comments refer only to the internal state
2. `ltu` is always updated to `block.timestamp`
3. Blue functions can be called by anyone
4. Purple functions can be called only by the sender

```mermaid
flowchart TD
    %% Functions
    ADJRPS([ADJUST_RPS]):::purple
    CR([CREATE]):::blue
    DP([DEPOSIT]):::blue
    PS([PAUSE]):::purple
    RFD([REFUND]):::purple
    RST([RESTART]):::purple
    WTD([WITHDRAW]):::blue

    %% Statuses
    NULL((NULL)):::grey
    STR((STREAMING)):::green
    PSD((PAUSED)):::yellow

    classDef grey fill:#b0b0b0,stroke:#333,stroke-width:2px;
    classDef green fill:#32cd32,stroke:#333,stroke-width:2px;
    classDef yellow fill:#ffff99,stroke:#333,stroke-width:2px;
    classDef blue fill:#99ccff,stroke:#333,stroke-width:2px;
    classDef purple fill:#D0CEE2,stroke:#333,stroke-width:2px;

    %% ltu is always updated to block.timestamp
    %% the "update" comments refer only to the internal state

    NULL --> CR
    CR -- "update rps\nupdate ltu" --> STR
    ADJRPS -- "update ltu\nupdate ra (+sa)\nupdate rps" -->  STR
    
    DP -- "update bal (+)" --> STR:::red
    DP -- "update bal (+)" --> PSD:::red
    
    RFD -- "update bal (-)" --> STR:::red
    RFD -- "update bal (-)" --> PSD:::red

    WTD -- "update ltu\nupdate bal (-)\nupdate ra (-)" --> STR
    WTD -- "update ltu\nupdate bal (-)\nupdate ra (-)" --> PSD

    STR --> PS
    PS -- "update ra (+sa)\nupdate rps (0)" --> PSD
    PSD --> RST
    RST -- "update ltu\nupdate rps" --> STR

    linkStyle 3,4,5,6,7,8 stroke:#ff0000,stroke-width:2px
```

---

### Streamed Amount

```mermaid
flowchart TD
sa([Streamed Amount - ra]):::green
res_sa(["rps*(block.timestamp - ltu)"]):::green
sa --> res_sa

classDef green fill:#32cd32,stroke:#333,stroke-width:2px;
```

### Withdrawable Amount

**Notes:** Debt greater than zero means:

1. `ra > bal`
2. `ra + sa > bal`

```mermaid
flowchart TD
    di0{ }:::blue0
    di1{ }:::blue0
    di2{ }:::blue0
    wa([Withdrawable Amount - wa]):::blue0
    res_0([0 ]):::blue1
    res_bal([bal]):::blue1
    res_ra([ra]):::blue1
    res_sum([sa + ra]):::blue1


    wa --> di0
    di0 -- "bal = 0" --> res_0
    di0 -- "bal > 0" --> di1
    di1 -- "debt > 0" --> res_bal
    di1 -- "debt = 0" --> di2
    di2 -- "isPaused" --> res_ra
    di2 -- "isNotPaused" --> res_sum

    classDef blue0 fill:#DAE8FC,stroke:#333,stroke-width:2px;
    classDef blue1 fill:#1BA1E2,stroke:#333,stroke-width:2px;
    linkStyle 1,2,3,4,5,6 stroke:#1BA1E2,stroke-width:2px
```

### Refundable Amount

```mermaid
    flowchart TD
    rfa([Refundable Amount - rfa]):::orange0
    res_rfa((bal - wa)):::orange1
    rfa --> res_rfa

    classDef orange0 fill:#FFA500,stroke:#333,stroke-width:2px;
    classDef orange1 fill:#FFCD28,stroke:#333,stroke-width:2px;

```

### Stream Debt

```mermaid
flowchart TD
    di0{ }:::red1
    sd([Stream Debt - sd]):::red0
    res_sd(("sa + ra - bal")):::red1
    res_zero([0]):::red1

    sd --> di0
    di0 -- "bal < sa + ra" --> res_sd
    di0 -- "bal >= sa + ra" --> res_zero

    classDef red0 fill:#EA6B66,stroke:#333,stroke-width:2px;
    classDef red1 fill:#FFCCCC,stroke:#333,stroke-width:2px;

```


</p>
</details> 


<details><summary>Drawio diagrams</summary>
<p>

### State diagram

![OE_state_diagram](https://github.com/sablier-labs/v2-open-ended/assets/99738872/51232e71-9e85-4b22-b4c6-e9f7254c00e9)

---

### Calculation diagrams

![Amounts_Calculation](https://github.com/sablier-labs/v2-open-ended/assets/99738872/cede8066-3c8b-4064-8aad-a38b95eff72a)


</p>
</details> 


Gist version:  https://gist.github.com/andreivladbrg/e5a0cf2120521e1d8ab429c35abcac14

**Reply by smol-ninja** [reply]
Great work on diagrams @andreivladbrg. Between these two, I prefer Drawio, the straight lines make it more organized and easier to understand. 

There are a suggestion on the state diagram:
1. `refund` should now be connected to both the states.

**Reply by andreivladbrg** [reply]
you are damn right, i missed this, will edit tomorrow

also, `deposit` https://github.com/sablier-labs/v2-open-ended/pull/113#discussion_r1612389170

**Reply by PaulRBerg** [reply]
"damn right" 😆

**Comment by andreivladbrg** [comment]
@smol-ninja since we have `DIAGRAMS.md` file should we close this?

**Reply by smol-ninja** [reply]
Yes.
